### ISSUE TITLE
ClassCastException with DefaultItemAnimator and adapter.set()

### ISSUE THREAD

**Issue by GreaseMonk** [mikepenz/FastAdapter#703]
## About this issue

- Briefly describe the issue

Setting an adapter repeatedly during a search with 3 types of classes of that implement IItem identical to RealmSampleUserItem, crashes the app due to animations in the DefaultItemAnimator (Please check the stack trace below).

Based on the following solution that i've found, I cannot resolve the issue that is directly caused from within the FastAdapter library. What happens is that the adapter is set again when an animation is in progress, and the onAnimationFinished callback in the RecyclerView class notifies the FastAdapter of this for a specific item index, but the item at that index is now of another IItem class. 
Setting `recyclerView.setItemAnimator(null);` resolves the issue.

A possible resolution would be to end all animations in the recyclerview when calling `adapter.set()`, but calling `recyclerView.clearAnimation();` before setting the new list does not resolve the issue.

With this ticket I hope to get more information and possibly other alternatives how to resolve the issue, or if this can be resolved from within the FastAdapter library.

- How can the issue be reproduced / sample code

1. Create 3 classes similar to RealmSampleUserItem
2. Set up a recyclerview with fast item adapter
3. Implement a search bar that will filter the items
4. Create a list of items (mixed classes) from step 1
5. Run the app and search.

Please note, when i search, i use `adapter.clear()` and then `adapter.set(filteredList)` instead of the adapter's filter function, because of filtering mechanics other than a string search.
Also, Before modifying the adapter or recyclerview, i am using `recyclerView.setLayoutFrozen(true/false)`. 

Crash log should show as follows:
[Open logcat log](https://pastebin.com/raw/07BpAxG9)

## Details
- [x] Used library version 3.2.5
- [x] Used support library version 27.1.1
- [x] Used gradle build tools version 3.1.3
- [x] Used tooling / Android Studio version (MacOS) Android Studio 3.1.3 #AI-173.4819257
- [x] Other used libraries, potential conflicting libraries (mike, I will PM you if you need this list)

## Checklist

- [x] Searched for [similar issues](https://github.com/mikepenz/FastAdapter/issues)
- [x] Checked out the [sample application](https://github.com/mikepenz/FastAdapter/tree/develop/app)
- [x] Read the [README](https://github.com/mikepenz/FastAdapter/blob/develop/README.md)
- [x] Checked out the [CHANGELOG](https://github.com/mikepenz/FastAdapter/releases)
- [x] Checked out the [MIGRATION GUIDE](https://github.com/mikepenz/FastAdapter/blob/develop/MIGRATION.md)

**Comment by mikepenz** [comment]
@GreaseMonk something seems off though. 

Can you please provide some sample code to help sorting this bug out?

**Comment by GreaseMonk** [comment]
[https://drive.google.com/file/d/148LaNyejL7-8rciR3EYXwmBUVPvS-bkD/view?usp=sharing](download ClassCastException project)

**Comment by mikepenz** [comment]
Thanks :). 

As discussion on hangouts there is a tiny sample bug in which most likely causes the crash in this sample. was not able to reproduce after solving it

**Comment by mikepenz** [comment]
So it seems in 

```java

    /**
     * is called in onViewDetachedFromWindow when the view is detached from the window
     *
     * @param viewHolder the viewHolder for the type at this position
     * @param position   the position of this viewHolder
     */
    @Override
    public void onViewDetachedFromWindow(RecyclerView.ViewHolder viewHolder, int position) {
        IItem item = FastAdapter.getHolderAdapterItem(viewHolder, position);
        if (item != null) {
            item.detachFromWindow(viewHolder);
            if (viewHolder instanceof FastAdapter.ViewHolder) {
                ((FastAdapter.ViewHolder) viewHolder).detachFromWindow(item);
            }
        }
    }
```

the `ViewHolder` does not match the position (retrieved from `getAdapterPosition` of the `VH`) anymore. so the adapter returns a different element for this case (position already has a different element). causing the cast for the `viewHolder` here: `item.detachFromWindow(viewHolder);` to fail

**Comment by mikepenz** [comment]
@GreaseMonk fixed in v3.3.0-rc02.

**Comment by mikepenz** [comment]
in addition did a release v3.2.9 with appcompat libs for you :)
